File: solution/055JumpGame/055JumpGame.go
package main

import "fmt"

func canJump(nums []int) bool {
	n := len(nums)
	i := 0
	for i < n-1 {
		if nums[i] == 0 {
			return false
		}
		k := i
		m := i
		for j := 1; j <= nums[i]; j++ {
			if i+j >= n-1 {
				return true
			} else if i+j+nums[i+j] > k {
				k = i + j + nums[i+j]
				m = i + j
			}
		}
		i = m
	}
	return i >= n-1
}

func main() {
	fmt.Printf("true  vs %v\n", canJump([]int{0}))
	fmt.Printf("true  vs %v\n", canJump([]int{1}))
	fmt.Printf("true  vs %v\n", canJump([]int{1, 2}))
	fmt.Printf("true  vs %v\n", canJump([]int{2, 0}))
	fmt.Printf("false vs %v\n", canJump([]int{1, 0, 2}))
	fmt.Printf("true  vs %v\n", canJump([]int{2, 3, 1, 1, 4}))
	fmt.Printf("false vs %v\n", canJump([]int{3, 2, 1, 0, 4}))
	fmt.Printf("true  vs %v\n", canJump([]int{1, 1, 2, 2, 0, 1, 1}))
	fmt.Printf("true  vs %v\n", canJump([]int{5, 9, 3, 2, 1, 0, 2, 3, 3, 1, 0, 0}))
}

File: src/055JumpGame/055JumpGame.rs
impl Solution {
    pub fn can_jump(nums: Vec<i32>) -> bool {
        if nums.len() == 0 {
            return false;
        }
        let mut pos = 0;
        while nums[pos] != 0 && (pos + nums[pos] as usize) < nums.len()-1 {
            let mut max_val = 0;
            let mut next = pos;
            for i in 1..=(nums[pos] as usize) {
                if (pos+i+nums[pos+i] as usize) >= max_val {
                    next = pos+i;
                    max_val = pos+i+nums[pos+i] as usize;
                }
            }
            pos = next;
        }
        return (pos + nums[pos] as usize) >= nums.len()-1;
    }
}
